### PR TITLE
Add indexer capacity, add historical boilerplate

### PR DESCRIPTION
**Pull request by lamcc21**
#50  #44

### CODE REVIEW

File: src/mappings/utils.ts [hunk 1]
```diff
@@ -288,6 +309,59 @@ export async function updateTotalDelegation(
...
+
+    const current =
+      stakeCurr?.mul(leverageLimit).sub(totalStakeCurr || 0) ||
+      BigNumber.from(0);
+    const after =
+      stakeAfter?.mul(leverageLimit).sub(totalStakeAfter || 0) ||
+      BigNumber.from(0);
+
+    const currentEra = await eraManager.eraNumber().then((r) => r.toNumber());
+
```
Comment (lamcc21) [inline src/mappings/utils.ts:348]: wasn't sure if to move this calculation to new function

File: src/mappings/staking.ts [hunk 2]
```diff
@@ -162,6 +157,10 @@ export async function handleRemoveDelegation(
   );
 
   await delegation.save();
+
+  if (source === indexer) {
```
Comment (mzxyz) [inline src/mappings/staking.ts]: why need to have this condition, if the source is delegator, we also need to update the indexer capacity

File: src/mappings/utils.ts [hunk 3]
```diff
@@ -288,6 +309,59 @@ export async function updateTotalDelegation(
...
+    new FrontierEthProvider()
+  );
+
+  const leverageLimit = await staking.indexerLeverageLimit();
+
+  if (indexer) {
+    const indexerStake = delegation?.amount;
+    const indexerTotalStake = indexer?.totalStake;
+
+    const stakeCurr = BigNumber.from(indexerStake?.value.value);
```
Comment (mzxyz) [inline src/mappings/utils.ts]: Please confirm that `BigNumber.from`  accept `undefined`. Seems it will throw error

**Comment by mzxyz** [comment]
Try to run it locally, it's easy to verify the capacity

File: src/mappings/utils.ts (hunk 3 above)
Comment (lamcc21) [reply src/mappings/utils.ts]: I don't think this condition would occur but I've updated for clarity.

File: src/mappings/utils.ts [hunk 4]
```diff
@@ -288,6 +309,61 @@ export async function updateTotalDelegation(
...
+    const stakeCurr = BigNumber.from(indexerStake?.value.value ?? 0);
+    const stakeAfter = BigNumber.from(indexerStake?.valueAfter.value ?? 0);
+
+    const totalStakeCurr = BigNumber.from(indexerTotalStake?.value.value ?? 0);
+    const totalStakeAfter = BigNumber.from(
+      indexerTotalStake?.valueAfter.value ?? 0
+    );
+
+    const current =
+      stakeCurr?.mul(leverageLimit).sub(totalStakeCurr || 0) ||
```
Comment (mzxyz) [inline src/mappings/utils.ts]: as you can have default value for `stakeCurr` etc, doesn't need to have `?.` and other default values in this expression.

Same with after

File: src/mappings/utils.ts [hunk 5]
```diff
@@ -288,6 +309,61 @@ export async function updateTotalDelegation(
...
+    new FrontierEthProvider()
+  );
+
+  const leverageLimit = await staking.indexerLeverageLimit();
+
+  if (indexer) {
+    const indexerStake = delegation?.amount;
+    const indexerTotalStake = indexer?.totalStake;
+
+    const stakeCurr = BigNumber.from(indexerStake?.value.value ?? 0);
```
Comment (mzxyz) [inline src/mappings/utils.ts]: Can have a helper function `bigNumberFrom(value: any): BigNumber`, when the value is undefined, can return BN(0).

**Comment by mzxyz** [comment]
Nice work, just 2 small things need to be updated, LGTM

**Review by mzxyz — APPROVED** [review]

**Pull request merged**